TrafficCapture/coreUtilities/src/main/java/org/opensearch/migrations/tracing/IInstrumentationAttributes.java

```diff
@@ -0,0 +1,109 @@
+package org.opensearch.migrations.tracing;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.metrics.DoubleHistogram;
+import io.opentelemetry.api.metrics.LongCounter;
+import io.opentelemetry.api.metrics.LongHistogram;
+import io.opentelemetry.api.metrics.LongUpDownCounter;
+import io.opentelemetry.api.trace.Span;
+import lombok.NonNull;
+
+import java.time.Duration;
+import java.util.ArrayList;
+
+public interface IInstrumentationAttributes {
+    AttributeKey<Boolean> HAD_EXCEPTION_KEY = AttributeKey.booleanKey("hadException");
+
+    IInstrumentationAttributes getEnclosingScope();
+
+    default Span getCurrentSpan() { return null; }
+
+    default AttributesBuilder fillAttributes(AttributesBuilder builder) {
+        return builder;
+    }
+
+    Throwable getObservedExceptionToIncludeInMetrics();
+
+    default @NonNull Attributes getPopulatedMetricAttributes(AttributesBuilder attributesBuilder) {
+        final var e = getObservedExceptionToIncludeInMetrics();
+        return e == null ? attributesBuilder.build() : attributesBuilder.put(HAD_EXCEPTION_KEY, true).build();
+    }
+
+    default Attributes getPopulatedSpanAttributes() {
+        return getPopulatedSpanAttributes(Attributes.builder());
+    }
+
+    default Attributes getPopulatedSpanAttributes(AttributesBuilder builder) {
+        return getPopulatedSpanAttributesBuilder(builder).build();
+    }
+
+    default AttributesBuilder getPopulatedSpanAttributesBuilder(AttributesBuilder builder) {
+        var currentObj = this;
+        var stack = new ArrayList<IInstrumentationAttributes>();
+        while (currentObj != null) {
+            stack.add(currentObj);
+            currentObj = currentObj.getEnclosingScope();
```

Comment:
It seems that a span is guaranteed to have all the attributes of it's parent from buildSpan definition, so is it the IInstrumentationAttributes job to loop through all the enclosingScopes to re-build these attributes?

Reply:
Yes.  It sounds like you're thinking that we could pull the enclosing scope's span value and not have to do the rebuilding.  That's probably correct for most of the code paths today, but one fuzzy spot would be that if an attribute was added to a span outside of fillAttributes (ie. getCurrentSpan().addAttribute() (e.g. adding an event or an exception).  Should we pick those up in downstream scopes?  Probably not.
However, should we be picking up parent scopes' attributes at all?  I really want to say "no", but early on in testing, I've noticed that if parent spans aren't closed out, only the children end up getting rendered by jaeger (and maybe exported by otel).  That leaves us in a bind to make children spans useful since they don't have any attributes.
I'd be interested in talking this through some.  It feels like a nuanced decision, though maybe one that doesn't matter too much.

Reply:
There should be inheritable values (like ids) and non-inheritable ones (like exceptions, sizes, and non-identifying runtime attributes).  I'll work on making that change.